component/component.go

```diff
@@ -190,9 +193,77 @@ func (f CreateDefaultConfigFunc) CreateDefaultConfig() Config {
 	return f()
 }
 
-// InstanceID uniquely identifies a component instance
+// pipelineDelim is the delimeter for internal representation of pipeline
+// component IDs.
+const pipelineDelim = byte(0x20)
```

Comment:
Unfortunately it turns out that IDs can contain a space :(

I mean I can define a receiver like this:

```yaml
receivers:
  "otlp/abc def":
    protocols:
      grpc:
        endpoint: localhost:4317
      http:
        endpoint: localhost:4318
```

Even a null byte in the name is accepted.

Seems a little lax to me. Perhaps revert to a map for now, but keep the method change, and we could follow up on making the type comparable later after finding a way to safely encode the IDs? Could either use size prefixing (non-breaking), or add some restrictions to IDs (breaking, but maybe a good idea anyway).

Reply:
We already introduced restrictions on the part before the `/` and had no user reports of it breaking them, so I would be in favor of doing the same for the name part (after the `/`)

Reply:
I can take a look at adding the same restrictions to the name as we currently have on the type. If it's not terrible, maybe we can make both changes? If it turns out to be more work than expected, we can keep the iterator, and use a map internally, then make InstanceID comparable when we have restrictions on the name.

Reply:
Yeah that makes sense to me

Reply:
I opened #10674 to handle restrictions to the name portion of a component.ID. The needs were slightly different so I introduced a different type for the name field.